### PR TITLE
Casualty selector breakup

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview

- Breaking up casualty selector to new classes
- Very few actual code changes, should be most reviewable commit-by-commit
- Added some new tests cases for casualty selector inner workings.

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[x] Code Cleanup or refactor
[] Configuration Change
[] Problem fix:  
[] Other:   

## Testing
- did some smoke testing while verifying various casualty scenarios on different maps

## Additional Review Notes

- please suggest any additional new test cases that would be really good to add to the two new casualty selector tests (f825494)

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLosses.java [hunk 1]
```diff
@@ -276,13 +297,13 @@ static void clearOolCache() {
...
+              + parameters.battlesite.getName()
               + "|"
-              + defending
+              + parameters.defending
               + "|"
-              + amphibious
+              + parameters.amphibious
               + "|"
               + targetsHashCode
               + "|"
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLosses.java:310]: I'm seeing this code snippet an odd amount of times.
Thoughts about extracting this to a helper method?

File: java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java [hunk 2]
```diff
@@ -30,6 +30,10 @@ public IntegerMap(final IntegerMap<T> integerMap) {
     this(integerMap.mapValues);
   }
 
+  public static <X> IntegerMap<X> of(final Map<X, Integer> map) {
+    return new IntegerMap<>(map);
+  }
+
```
Comment (RoiEXLab) [inline java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java:36]: What is this method for and why isn't the constructor enough?

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLosses.java (hunk 1 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLosses.java:310]: There are more areas of copy / paste than this too, code climate does a pretty decent job pointing them out.

This update is just preliminary, intentionally not modifying too much of the existing code yet. It needs to be done iteratively. I'm hoping as well that this hashing function can be removed as part of one those iterations.

File: java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java (hunk 2 above)
Comment (DanVanAtta) [reply java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java:36]: Convenience and avoids the diamond notation:
eg:
```
new IntegerMap<>(Map.of(...));
```
vs:
```
IntegerMap.of(Map.of(...))
```

Guava does that type of think all over the place, type inference in java is much smarter about static calls than it is about `new`

Ultimately I think `IntegerMap` is possibly a pre-Java 1.5 artifact from before there were even generics, we should consider replacing `IntegerMap` with `Map<T, Integer>`.

**Comment by DanVanAtta** [comment]
@RoiEXLab did you happen to spot any interesting test cases that could be added?

@ron-murhammer curious if you had a chance to look at the tests in this PR and could recommend adding additional interesting OOL tests.

**Comment by ron-murhammer** [comment]
@DanVanAtta Anything with multiple different support attachments that aren't just +1 to 1 unit would be good. I think 270bc elephants are a good example of that. Also something with subs/destroyers where strength is the same between taking a destroyer and non-destroyer against subs and it should take the non-destroyer.

File: java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java (hunk 2 above)
Comment (RoiEXLab) [reply java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java:36]: @DanVanAtta I should've clarified:
I don't think I've seen this method being used inside of this PR.
So I suspect it's unused

Comment (DanVanAtta) [reply java-extras/src/main/java/org/triplea/java/collections/IntegerMap.java:36]: FWIW - test cases make use of it:
https://github.com/triplea-game/triplea/pull/6276/files#diff-98b2c58d739b161b3aa5cc607fe5b6baR42

**Pull request merged**